### PR TITLE
Rewrite as WebExtension

### LINKED ISSUE #18: Electrolysis disabled by this plugin
When "open tabs next to current" is enabled, Electrolysis (Multi-Process Windows in about:support) is disabled.

### PR DESCRIPTION
**Pull request by samiljan**
Hi there,
I gave rewriting this as a WebExtension a try, since this was my only not multiprocess-compatible extension. This resolves #18.

The basic functionality was pretty straight forward to implement, but I couldn't manage to keep the keybinding for opening in the default location the same, or differentiate between creating a new tab from keyboard or the + button in the tab bar with the limitations of WebExtensions. 

Since it loses some functionality maybe it would be best to publish this as a new extension? What do you think?

regards,
Sam

### CODE REVIEW

**Comment by sblask** [comment]
Thanks for this! I'd merge it, but as you say, it breaks functionality. Most notably the ordering after a session restore - I assume at least as there is no code to handle it. This and the other problems I don't actually care about, but people asked me to implement them. I have no hope that these problems will be fixable though as I don't see any API for them coming. You also do not move tabs opened in the background. That would be a problem for me and if you'd want to keep it, that should be an option.

Creating a new extension is an option - if you want to maintain it. I'd actually use it myself instead of this one if it works the way I expect (I only created this one, because I was missing the functionality). If multiprocessing is your only issue, that should fix itself with Firefox 53, where it will be enabled unless add-ons explicitly disable multiprocessing.

File: .eslintrc [hunk 1]
```diff
@@ -1,17 +1,16 @@
 {
+    "globals": {
```
Comment (sblask) [inline .eslintrc]: There is actually a webextension env that you can use (like here: https://github.com/sblask/firefox-simple-form-fill/blob/master/.eslintrc)

File: index.js [hunk 2]
```diff
@@ -1,65 +1,62 @@
-const core = require("sdk/view/core");
-const hotkeys = require("sdk/hotkeys");
-const tabs = require("sdk/tabs");
-const windows = require("sdk/windows").browserWindows;
 
-const helpers = require("./lib/helpers");
-const state = require("./lib/state");
+function activatePlugin(initialActiveTabs) {
+    const lastActiveTab = new Map(initialActiveTabs.map(t => [t.windowId, t.id]));
+    const currentActiveTab = new Map(lastActiveTab);
```
Comment (sblask) [inline index.js]: `currentActiveTab` isn't used?

File: index.js [hunk 3]
```diff
@@ -1,65 +1,62 @@
...
+    browser.tabs.onActivated.addListener(({tabId, windowId}) => {
+        lastActiveTab.set(windowId, currentActiveTab.get(windowId));
+        currentActiveTab.set(windowId, tabId)
+        //console.log("onActivated", {tabId, windowId, lastActiveTab: lastActiveTab.get(windowId)});
     });
 
-    tabs.on("open", helpers.maybeMoveTab);
+    function getNextToCurrentTabIndex(windowId) {
+        const last = lastActiveTab.get(windowId);
+        if(last === undefined) {
```
Comment (sblask) [inline index.js]: This shouldn't actually happen?

File: index.js [hunk 4]
```diff
@@ -1,65 +1,62 @@
...
+            }).catch(e => {
+                if(e !== "no active tab") {
+                    throw e;
+                }
+            });
     }
 
-    windows.on("open", registerListeners);
-};
+    let enabled = true;
```
Comment (sblask) [inline index.js]: You could also add/remove `tabs.onCreated` listener in `onCommand` and get rid of this?

File: .eslintrc (hunk 1 above)
Comment (samiljan) [inline .eslintrc]: Nice!

File: index.js (hunk 2 above)
Comment (samiljan) [inline index.js]: It's used to set lastActiveTab when the active tab changes

File: index.js (hunk 3 above)
Comment (samiljan) [inline index.js]: It happens after opening new windows... tabs.onCreated is triggered before tabs.onActivated. I could probably avoid it if I handle windows.onCreated though

File: index.js (hunk 4 above)
Comment (samiljan) [inline index.js]: Ah, yes, that could be a better way :)

**Comment by samiljan** [comment]
Thanks for the feedback :)
I disabled background tab positioning since it didn't work well - I think it can be handled too but I haven't thought it through yet. I've noticed "undo close tab" get positioned weirdly too.

It seems that session restore is actually working fine now, since the active tab is created first in a restored window, and the other tabs are background tabs that are not currently positioned. It can probably be improved too. I'll try to take a look at it soon!

**Comment by DaAwesomeP** [comment]
I'm glad to see that this addon will survive (or at least mostly) the *Great Addon Purge.*

It is possible to remove the default new tab button from the toolbar via the customize toolbars page, so adding a dedicated button to create a new tab via the addon would be better than nothing. If <kbd>Ctrl</kbd>+ <kbd>T</kbd> can't be overriden, then it simply can't be overriden.

**Comment by DaAwesomeP** [comment]
@samiljan I'm using your fork currently, and I can't detect any issues with it. <kbd>Ctrl</kbd> + <kbd>T</kbd> and the new tab button seem to work just fine. I'm not sure what the option about repositioning the tabs does though.

**Comment by samiljan** [comment]
@DaAwesomeP The setting mostly affects "Undo Close Tab". If you enable the setting, restored tabs are positioned where they were when they were closed, like they normally are. But it also means that if you open multiple tabs from links in one tab, the later tabs will be positioned further to the right - and that was not the case using the old extension.

It seems the WebExtension way of repositioning tabs works a bit worse in Firefox 53... Now the newly opened tab will be scrolled to the far left of the tab bar if you have more tabs than can be shown at the same time - very annoying! And I think my hacky way of detecting a session restore fails sometimes, messing up the order of the tabs. I really hope the WebExtensions API will be extended with more events and event details related to tabs, such as the "source" of the new tab.

**Comment by DaAwesomeP** [comment]
> It seems the WebExtension way of repositioning tabs works a bit worse in Firefox 53... Now the newly opened tab will be scrolled to the far left of the tab bar if you have more tabs than can be shown at the same time - very annoying!

I'm in Nightly (FF55), and the scrolling bit is definitely very annoying, but I guess there's no way around that.

**Comment by samiljan** [comment]
> the scrolling bit is definitely very annoying, I guess there's no way around that.
No, it doesn't seem like there is. What is really needed is an API for deciding where to put a new tab.
It seems there is a plan to extend the tabs api: https://bugzilla.mozilla.org/show_bug.cgi?id=1333837#c9

**Comment by sblask** [comment]
I started working on a port taking the ideas from your implementation. It's still missing the extra shortcut and I won't do the hacky session recovery detection, but I think it should work and I managed to keep it much shorter. See #22

**Pull request closed without merging**